src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -227,12 +289,56 @@ public virtual string CreateToken(string payload, SigningCredentials signingCred
+        /// <exception cref="ArgumentNullException">if <paramref name="compressionAlgorithm"/> is null.</exception>
+        /// <exception cref="ArgumentNullException">if <paramref name="additionalHeaderClaims"/> is null.</exception>
+        /// <returns>A JWE in compact serialization format.</returns>
+        public virtual string CreateToken(string payload, SigningCredentials signingCredentials, EncryptingCredentials encryptingCredentials, string compressionAlgorithm, IDictionary<string, object> additionalHeaderClaims)
+        {
+            if (string.IsNullOrEmpty(payload))
+                throw LogHelper.LogArgumentNullException(nameof(payload));
+
+            if (signingCredentials == null)
+                throw LogHelper.LogArgumentNullException(nameof(signingCredentials));
+
+            if (encryptingCredentials == null)
+                throw LogHelper.LogArgumentNullException(nameof(encryptingCredentials));
+
+            if (string.IsNullOrEmpty(compressionAlgorithm))
+                throw LogHelper.LogArgumentNullException(nameof(compressionAlgorithm));
+
+            if (additionalHeaderClaims == null)
+                throw LogHelper.LogArgumentNullException(nameof(additionalHeaderClaims));
+
+            return CreateTokenPrivate(JObject.Parse(payload), signingCredentials, encryptingCredentials, compressionAlgorithm, additionalHeaderClaims);
         }
 
-        private string CreateTokenPrivate(JObject payload, SigningCredentials signingCredentials, EncryptingCredentials encryptingCredentials, string algorithm)
+        private string CreateTokenPrivate(JObject payload, SigningCredentials signingCredentials, EncryptingCredentials encryptingCredentials, string algorithm, IDictionary<string, object> additionalHeaderClaims)
         {
-            if (!JsonWebTokenManager.KeyToHeaderCache.TryGetValue(JsonWebTokenManager.GetHeaderCacheKey(signingCredentials), out string rawHeader))
+
+            if (additionalHeaderClaims != null && (additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.Alg)
+                || additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.Kid)
+                || additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.Typ)
+                || additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.X5t)
+                || additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.Enc)
+                || additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.Zip)))
+                throw LogHelper.LogExceptionMessage(new SecurityTokenException(LogHelper.FormatInvariant(LogMessages.IDX14116, nameof(additionalHeaderClaims))));
+
+            // if 'encryptingCredentials' != null it means additionalClaims should be added to the outer JWT header.
+            if (encryptingCredentials == null && additionalHeaderClaims != null || !JsonWebTokenManager.KeyToHeaderCache.TryGetValue(JsonWebTokenManager.GetHeaderCacheKey(signingCredentials), out string rawHeader))
```

Comment:
I don't think that we should ever cache headers when _additionalHeaderClaims_ is not null.
A header cache key is formed out of a security key type, key id, and an algorithm.

Image that the first call has a key=="1" and alg == "RS256" and some additional header claims.
As header cache is empty at that point we will cache the header, together with the additional claims.

If we have the second call with the same key id and alg ("1", "RS256") but without additional header claims, we will actually use the header cached in the first call (with additional claims).

Reply:
You're right; this was an oversight on my part. Fixed this so that we never cache a header that had additional claims added to it.

Also added in tests for this.